src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -6272,20 +6272,28 @@ public static Task<TResult[]> WhenAll<TResult>(IEnumerable<Task<TResult>> tasks)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.tasks);
             }
 
-            List<Task<TResult>> taskList = new List<Task<TResult>>();
-            foreach (Task<TResult> task in tasks)
+            using var enumerator = tasks.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                return new Task<TResult[]>(false, [], TaskCreationOptions.None, default);
+            }
+
+            var taskList = new List<Task<TResult>>();
```

Comment:
The code is still not as optimal as it can. The non-result version `public static Task WhenAll(IEnumerable<Task> tasks)` has better optimization pattern.

Reply:
I experimented locally with using ValueListBuilder; in some cases it's much better, in some cases, it's worse.